### ISSUE TITLE
Proposed fix to allow DynamicFont kerning to be re-enabled

### ISSUE THREAD

**Issue by follower** [godotengine/godot#38731]
----

### TL;DR: The Bug & Cause

 * `DynamicFont` kerning was removed by https://github.com/godotengine/godot/pull/24777 due to issues with rendering some fonts as described in https://github.com/godotengine/godot/issues/21965.

 * The underlying issue with the removed kerning-related code was that the call to `FT_Get_Kerning()` supplied arguments that were *character codes* rather than *glyph indexes*.

 *  The observed impact of the issue was that the kerning value applied was intended for a different pair of characters than the pair displayed.

    For certain font & character combinations the difference between the correct & incorrect kerning values was significant. e.g. for `PetitLatin.ttf` the kerning for `t|` ("tee pipe") was used to kern `ck` which resulted in character overlap.

### TL;DR: The Fix

Potential fix steps:

* Revert https://github.com/godotengine/godot/commit/a9b7843dca3ea3eda9f9f890a3e2c78542ca7561

* Apply this patch:

```diff
diff --git a/scene/resources/dynamic_font.cpp b/scene/resources/dynamic_font.cpp
index e165f7c34b..be6d6b517f 100644
--- a/scene/resources/dynamic_font.cpp
+++ b/scene/resources/dynamic_font.cpp
@@ -282,7 +282,7 @@ float DynamicFontAtSize::_get_kerning_advance(const DynamicFontAtSize *font, Cha
 
 	if (p_next) {
 		FT_Vector delta;
-		FT_Get_Kerning(font->face, p_char, p_next, FT_KERNING_DEFAULT, &delta);
+		FT_Get_Kerning(font->face, FT_Get_Char_Index(font->face, p_char), FT_Get_Char_Index(font->face, p_next), FT_KERNING_DEFAULT, &delta);
 		advance = (delta.x / 64.0) / oversampling;
 	}
```

This fix would both re-enable kerning[1] for `DynamicFont` and fix the previously observed kerning issue.

[1] For fonts with `kern` tables but still not for `GPOS`/`kern` tables.

### TL;DR: Pretty pictures

Comparison screenshots based on `HEAD` 3.2 branch yesterday:

<details>

 * The bug (incorrect kerning applied):

    ![buggy-kern-tee-pipe-cropped](https://user-images.githubusercontent.com/189962/81881818-41ab0d80-95e5-11ea-8bfb-87b77202fa1c.png)

 * Current situation (no kerning applied):

    ![current--no-kern--tee-pipe--cropped](https://user-images.githubusercontent.com/189962/81881832-4e2f6600-95e5-11ea-8414-76a063f49b16.png)

 * After fix applied (correct kerning applied):

    ![good--kern--tee-pipe--fixed--cropped](https://user-images.githubusercontent.com/189962/81881855-5e474580-95e5-11ea-8ca7-1ea98e5ed33d.png)

    Note the correct `c`+`k` kerning and kerning that reduces the gap for the `u`+`c` & `t`+`|` character pairs.

</details>

----

*[WIP: Intend to add additional details & code references.]*

**Comment by Anutrix** [comment]
Can make a PR?

**Comment by follower** [comment]
@Anutrix Yes, although I'm currently knee deep in yaks. :)

**Comment by pbgamestudio** [comment]
Hi, just wondering if there is progress in re-adding kerning to DynamicFonts? :)

**Comment by akien-mga** [comment]
Fixed by #49377.

**Comment by follower** [comment]
Can confirm that as of Godot 3.4 beta 2 the official "ducky tee-pipe"(TM) kerning test now appears to pass. :)

![ducky--tee-pipe--test--Screenshot from 2021-07-28 16 36 49](https://user-images.githubusercontent.com/189962/127264755-e9d67d80-76f5-4709-90d6-6dc2fc68d7c4.png)

Font file from [Petit Latin Font](https://www.dafont.com/petit-latin.font?text=ducky+"tee-pipe"+t|).
